compiler/optimizer/OMRSimplifier.hpp

```diff
@@ -121,7 +122,127 @@ class Simplifier : public TR::Optimization
    void prepareToReplaceNode(TR::Node * node);
 
    void anchorOrderDependentNodesInSubtree(TR::Node *node, TR::Node *replacement, TR::TreeTop* anchorTree);
-   TR::Node * unaryCancelOutWithChild(TR::Node * node, TR::Node * firstChild, TR::TreeTop* anchorTree, TR::ILOpCodes opcode, bool anchorChildren=true);
+
+   virtual bool isLegalToUnaryCancel(TR::Node *node, TR::Node *firstChild, TR::ILOpCodes opcode) 
+      { return true; }
+   virtual TR::Node *unaryCancelOutWithChild(TR::Node *node, TR::Node *firstChild, TR::TreeTop *anchorTree, TR::ILOpCodes opcode, bool anchorChildren=true);
+
+   /**
+    * Checks IL tree patterns to see if folding can be done safely.
+    * 
+    * @param[in] node The tree node that we are trying to fold
+    * @param[in] firstChild The first child of node
+    * @return Returns true if node can be folded; false otherwise
+    */
+   virtual bool isLegalToFold(TR::Node *node, TR::Node *firstChild) 
+      { return true; }
+
+   /**
+    * Checks if the bound value is great than or equal to the length value.
+    *
+    * @param[in] boundChild The tree node corresponding to the bound value
+    * @param[in] lengthChild The tree node corresponding to the length value
+    * @return Returns true if bound is great than or equal to length; 
+    * false otherwise
+    */
+   virtual bool isLengthGEBound(TR::Node *boundChild, TR::Node *lengthChild);
```

Comment:
Good point, the name has been changed to isBoundDefinitelyGELength. I agree with you that the comparison currently has nothing to do with lengths and bounds, but a method that compares its two arguments might add other IL tree patterns in the future that might not apply here so I would like to leave the name with its "tie" to array bound checking.